Let binding assertion failure

Boogie crashes on the following input:

```
type AAA;
type BBB;
type CCC;

function Apple(AAA, BBB): CCC;
function Banana(int): BBB;

procedure Proc()
{
  var m: [AAA]CCC;
  m :=
      (lambda aaa: AAA :: 
          (var g := 15;
             Apple(aaa, Banana(700))
          )
      );
}
```

Here is the output:

```
$ boogie boogie-crash-small.bpl
Boogie program verifier version 2.4.1.10503, Copyright (c) 2003-2014, Microsoft.
[ERROR] FATAL UNHANDLED EXCEPTION: System.AggregateException: One or more errors occurred. ---> System.Diagnostics.Contracts.ContractException: Assertion failed.
  at System.Runtime.CompilerServices.ContractHelper.TriggerFailureImplementation (System.Diagnostics.Contracts.ContractFailureKind kind, System.String displayMessage, System.String userMessage, System.String conditionText, System.Exception innerException) [0x0002f] in <bb7b695b8c6246b3ac1646577aea7650>:0
  at System.Runtime.CompilerServices.ContractHelper.TriggerFailure (System.Diagnostics.Contracts.ContractFailureKind kind, System.String displayMessage, System.String userMessage, System.String conditionText, System.Exception innerException) [0x00000] in <bb7b695b8c6246b3ac1646577aea7650>:0
  at System.Diagnostics.Contracts.Contract.ReportFailure (System.Diagnostics.Contracts.ContractFailureKind failureKind, System.String userMessage, System.String conditionText, System.Exception innerException) [0x0003a] in <bb7b695b8c6246b3ac1646577aea7650>:0
  at System.Diagnostics.Contracts.Contract.Assert (System.Boolean condition) [0x00003] in <bb7b695b8c6246b3ac1646577aea7650>:0
  at Microsoft.Boogie.VCExprAST.VCExprLetBinding..ctor (Microsoft.Boogie.VCExprAST.VCExprVar v, Microsoft.Boogie.VCExprAST.VCExpr e) [0x00027] in <b410fe19c07d4326b5592482c71fc76c>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.VisitLetExpr (Microsoft.Boogie.LetExpr node) [0x000de] in <b410fe19c07d4326b5592482c71fc76c>:0
  at Microsoft.Boogie.LetExpr.StdDispatch (Microsoft.Boogie.StandardVisitor visitor) [0x00001] in <ecbf282659c24407b75625b1ae5a3398>:0
  at Microsoft.Boogie.ReadOnlyVisitor.Visit (Microsoft.Boogie.Absy node) [0x00001] in <ecbf282659c24407b75625b1ae5a3398>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.Translate (Microsoft.Boogie.Expr expr) [0x00001] in <b410fe19c07d4326b5592482c71fc76c>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.TranslateNAryExpr (Microsoft.Boogie.NAryExpr node) [0x001b6] in <b410fe19c07d4326b5592482c71fc76c>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.VisitNAryExpr (Microsoft.Boogie.NAryExpr node) [0x00001] in <b410fe19c07d4326b5592482c71fc76c>:0
  at Microsoft.Boogie.NAryExpr.StdDispatch (Microsoft.Boogie.StandardVisitor visitor) [0x00001] in <ecbf282659c24407b75625b1ae5a3398>:0
  at Microsoft.Boogie.ReadOnlyVisitor.Visit (Microsoft.Boogie.Absy node) [0x00001] in <ecbf282659c24407b75625b1ae5a3398>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.Translate (Microsoft.Boogie.Expr expr) [0x00001] in <b410fe19c07d4326b5592482c71fc76c>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.TranslateQuantifierExpr (Microsoft.Boogie.QuantifierExpr node) [0x000b2] in <b410fe19c07d4326b5592482c71fc76c>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.VisitQuantifierExpr (Microsoft.Boogie.QuantifierExpr node) [0x00001] in <b410fe19c07d4326b5592482c71fc76c>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.VisitForallExpr (Microsoft.Boogie.ForallExpr node) [0x00001] in <b410fe19c07d4326b5592482c71fc76c>:0
  at Microsoft.Boogie.ForallExpr.StdDispatch (Microsoft.Boogie.StandardVisitor visitor) [0x00001] in <ecbf282659c24407b75625b1ae5a3398>:0
  at Microsoft.Boogie.ReadOnlyVisitor.Visit (Microsoft.Boogie.Absy node) [0x00001] in <ecbf282659c24407b75625b1ae5a3398>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.Translate (Microsoft.Boogie.Expr expr) [0x00001] in <b410fe19c07d4326b5592482c71fc76c>:0
  at Microsoft.Boogie.DeclFreeProverContext.AddAxiom (Microsoft.Boogie.Axiom ax, System.String attributes) [0x0001c] in <28af89544a894d3aba7d17408628decf>:0
  at Microsoft.Boogie.Checker.Setup (Microsoft.Boogie.Program prog, Microsoft.Boogie.ProverContext ctx) [0x000d5] in <28af89544a894d3aba7d17408628decf>:0
  at Microsoft.Boogie.Checker..ctor (VC.ConditionGeneration vcgen, Microsoft.Boogie.Program prog, System.String logFilePath, System.Boolean appendLogFile, System.Int32 timeout, System.Int32 rlimit, Microsoft.Boogie.ProverContext ctx) [0x0013d] in <28af89544a894d3aba7d17408628decf>:0
  at VC.ConditionGeneration.FindCheckerFor (System.Int32 timeout, System.Int32 rlimit, System.Boolean isBlocking, System.Int32 waitTimeinMs, System.Int32 maxRetries) [0x001aa] in <28af89544a894d3aba7d17408628decf>:0
  at VC.VCGen.VerifyImplementation (Microsoft.Boogie.Implementation impl, Microsoft.Boogie.VerifierCallback callback) [0x00610] in <28af89544a894d3aba7d17408628decf>:0
  at VC.ConditionGeneration.VerifyImplementation (Microsoft.Boogie.Implementation impl, System.Collections.Generic.List`1[Microsoft.Boogie.Counterexample]& errors, System.String requestId) [0x00019] in <28af89544a894d3aba7d17408628decf>:0
  at Microsoft.Boogie.ExecutionEngine.VerifyImplementation (Microsoft.Boogie.Program program, Microsoft.Boogie.PipelineStatistics stats, Microsoft.Boogie.ErrorReporterDelegate er, System.String requestId, System.Collections.Generic.Dictionary`2[TKey,TValue] extractLoopMappingInfo, Microsoft.Boogie.Implementation[] stablePrioritizedImpls, System.Int32 index, Microsoft.Boogie.ExecutionEngine+OutputCollector outputCollector, System.Collections.Generic.List`1[T] checkers, System.String programId) [0x00243] in <e5643ff9bd264395ad10c8ba7b71764e>:0
  at Microsoft.Boogie.ExecutionEngine+<>c__DisplayClass27_2.<InferAndVerify>b__5 (System.Object dummy) [0x00056] in <e5643ff9bd264395ad10c8ba7b71764e>:0
  at System.Threading.Tasks.Task.InnerInvoke () [0x00025] in <bb7b695b8c6246b3ac1646577aea7650>:0
  at System.Threading.Tasks.Task.Execute () [0x00010] in <bb7b695b8c6246b3ac1646577aea7650>:0
   --- End of inner exception stack trace ---
  at System.AggregateException.Handle (System.Func`2[T,TResult] predicate) [0x00064] in <bb7b695b8c6246b3ac1646577aea7650>:0
  at Microsoft.Boogie.ExecutionEngine.InferAndVerify (Microsoft.Boogie.Program program, Microsoft.Boogie.PipelineStatistics stats, System.String programId, Microsoft.Boogie.ErrorReporterDelegate er, System.String requestId) [0x00523] in <e5643ff9bd264395ad10c8ba7b71764e>:0
  at Microsoft.Boogie.ExecutionEngine.ProcessFiles (System.Collections.Generic.List`1[T] fileNames, System.Boolean lookForSnapshots, System.String programId) [0x002b7] in <e5643ff9bd264395ad10c8ba7b71764e>:0
  at Microsoft.Boogie.OnlyBoogie.Main (System.String[] args) [0x0029d] in <7fe34dbd3a6f4d8387ecb00830f06845>:0
---> (Inner Exception #0) System.Diagnostics.Contracts.ContractException: Assertion failed.
  at System.Runtime.CompilerServices.ContractHelper.TriggerFailureImplementation (System.Diagnostics.Contracts.ContractFailureKind kind, System.String displayMessage, System.String userMessage, System.String conditionText, System.Exception innerException) [0x0002f] in <bb7b695b8c6246b3ac1646577aea7650>:0
  at System.Runtime.CompilerServices.ContractHelper.TriggerFailure (System.Diagnostics.Contracts.ContractFailureKind kind, System.String displayMessage, System.String userMessage, System.String conditionText, System.Exception innerException) [0x00000] in <bb7b695b8c6246b3ac1646577aea7650>:0
  at System.Diagnostics.Contracts.Contract.ReportFailure (System.Diagnostics.Contracts.ContractFailureKind failureKind, System.String userMessage, System.String conditionText, System.Exception innerException) [0x0003a] in <bb7b695b8c6246b3ac1646577aea7650>:0
  at System.Diagnostics.Contracts.Contract.Assert (System.Boolean condition) [0x00003] in <bb7b695b8c6246b3ac1646577aea7650>:0
  at Microsoft.Boogie.VCExprAST.VCExprLetBinding..ctor (Microsoft.Boogie.VCExprAST.VCExprVar v, Microsoft.Boogie.VCExprAST.VCExpr e) [0x00027] in <b410fe19c07d4326b5592482c71fc76c>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.VisitLetExpr (Microsoft.Boogie.LetExpr node) [0x000de] in <b410fe19c07d4326b5592482c71fc76c>:0
  at Microsoft.Boogie.LetExpr.StdDispatch (Microsoft.Boogie.StandardVisitor visitor) [0x00001] in <ecbf282659c24407b75625b1ae5a3398>:0
  at Microsoft.Boogie.ReadOnlyVisitor.Visit (Microsoft.Boogie.Absy node) [0x00001] in <ecbf282659c24407b75625b1ae5a3398>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.Translate (Microsoft.Boogie.Expr expr) [0x00001] in <b410fe19c07d4326b5592482c71fc76c>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.TranslateNAryExpr (Microsoft.Boogie.NAryExpr node) [0x001b6] in <b410fe19c07d4326b5592482c71fc76c>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.VisitNAryExpr (Microsoft.Boogie.NAryExpr node) [0x00001] in <b410fe19c07d4326b5592482c71fc76c>:0
  at Microsoft.Boogie.NAryExpr.StdDispatch (Microsoft.Boogie.StandardVisitor visitor) [0x00001] in <ecbf282659c24407b75625b1ae5a3398>:0
  at Microsoft.Boogie.ReadOnlyVisitor.Visit (Microsoft.Boogie.Absy node) [0x00001] in <ecbf282659c24407b75625b1ae5a3398>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.Translate (Microsoft.Boogie.Expr expr) [0x00001] in <b410fe19c07d4326b5592482c71fc76c>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.TranslateQuantifierExpr (Microsoft.Boogie.QuantifierExpr node) [0x000b2] in <b410fe19c07d4326b5592482c71fc76c>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.VisitQuantifierExpr (Microsoft.Boogie.QuantifierExpr node) [0x00001] in <b410fe19c07d4326b5592482c71fc76c>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.VisitForallExpr (Microsoft.Boogie.ForallExpr node) [0x00001] in <b410fe19c07d4326b5592482c71fc76c>:0
  at Microsoft.Boogie.ForallExpr.StdDispatch (Microsoft.Boogie.StandardVisitor visitor) [0x00001] in <ecbf282659c24407b75625b1ae5a3398>:0
  at Microsoft.Boogie.ReadOnlyVisitor.Visit (Microsoft.Boogie.Absy node) [0x00001] in <ecbf282659c24407b75625b1ae5a3398>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.Translate (Microsoft.Boogie.Expr expr) [0x00001] in <b410fe19c07d4326b5592482c71fc76c>:0
  at Microsoft.Boogie.DeclFreeProverContext.AddAxiom (Microsoft.Boogie.Axiom ax, System.String attributes) [0x0001c] in <28af89544a894d3aba7d17408628decf>:0
  at Microsoft.Boogie.Checker.Setup (Microsoft.Boogie.Program prog, Microsoft.Boogie.ProverContext ctx) [0x000d5] in <28af89544a894d3aba7d17408628decf>:0
  at Microsoft.Boogie.Checker..ctor (VC.ConditionGeneration vcgen, Microsoft.Boogie.Program prog, System.String logFilePath, System.Boolean appendLogFile, System.Int32 timeout, System.Int32 rlimit, Microsoft.Boogie.ProverContext ctx) [0x0013d] in <28af89544a894d3aba7d17408628decf>:0
  at VC.ConditionGeneration.FindCheckerFor (System.Int32 timeout, System.Int32 rlimit, System.Boolean isBlocking, System.Int32 waitTimeinMs, System.Int32 maxRetries) [0x001aa] in <28af89544a894d3aba7d17408628decf>:0
  at VC.VCGen.VerifyImplementation (Microsoft.Boogie.Implementation impl, Microsoft.Boogie.VerifierCallback callback) [0x00610] in <28af89544a894d3aba7d17408628decf>:0
  at VC.ConditionGeneration.VerifyImplementation (Microsoft.Boogie.Implementation impl, System.Collections.Generic.List`1[Microsoft.Boogie.Counterexample]& errors, System.String requestId) [0x00019] in <28af89544a894d3aba7d17408628decf>:0
  at Microsoft.Boogie.ExecutionEngine.VerifyImplementation (Microsoft.Boogie.Program program, Microsoft.Boogie.PipelineStatistics stats, Microsoft.Boogie.ErrorReporterDelegate er, System.String requestId, System.Collections.Generic.Dictionary`2[TKey,TValue] extractLoopMappingInfo, Microsoft.Boogie.Implementation[] stablePrioritizedImpls, System.Int32 index, Microsoft.Boogie.ExecutionEngine+OutputCollector outputCollector, System.Collections.Generic.List`1[T] checkers, System.String programId) [0x00243] in <e5643ff9bd264395ad10c8ba7b71764e>:0
  at Microsoft.Boogie.ExecutionEngine+<>c__DisplayClass27_2.<InferAndVerify>b__5 (System.Object dummy) [0x00056] in <e5643ff9bd264395ad10c8ba7b71764e>:0
  at System.Threading.Tasks.Task.InnerInvoke () [0x00025] in <bb7b695b8c6246b3ac1646577aea7650>:0
  at System.Threading.Tasks.Task.Execute () [0x00010] in <bb7b695b8c6246b3ac1646577aea7650>:0 <---
```

--- a/Source/Core/AbsyQuant.cs
+++ b/Source/Core/AbsyQuant.cs
@@ -1119,11 +1119,16 @@ namespace Microsoft.Boogie {
     public override void ComputeFreeVariables(Set freeVars) {
       //Contract.Requires(freeVars != null);
 
+      foreach (var e in Rhss) {
+        e.ComputeFreeVariables(freeVars);
+      }
       foreach (var v in Dummies) {
         Contract.Assert(v != null);
         Contract.Assert(!freeVars[v]);
       }
-      Body.ComputeFreeVariables(freeVars);
+      foreach (var v in Dummies) {
+        freeVars.AddRange(v.TypedIdent.Type.FreeVariables);
+      }
       for (var a = Attributes; a != null; a = a.Next) {
         foreach (var o in a.Params) {
           var e = o as Expr;
@@ -1132,13 +1137,8 @@ namespace Microsoft.Boogie {
           }
         }
       }
-      foreach (var v in Dummies) {
-        freeVars.AddRange(v.TypedIdent.Type.FreeVariables);
-      }
+      Body.ComputeFreeVariables(freeVars);
       freeVars.RemoveRange(Dummies);
-      foreach (var e in Rhss) {
-        e.ComputeFreeVariables(freeVars);
-      }
     }
   }
 }

--- a/Source/Core/MaxHolesLambdaLifter.cs
+++ b/Source/Core/MaxHolesLambdaLifter.cs
@@ -131,8 +131,7 @@ class MaxHolesLambdaLifter : StandardVisitor {
 
             replacements.AddRange(bodyTemplate.GetReplacements());
             _templates[node] = new TemplateWithBoundVariables(replacements);
-        }
-        else {
+        } else {
             var newRhss = from arg in varBodies select ((TemplateNoBoundVariables) _templates[arg]).GetReplacement();
             LambdaLiftingTemplate template = new TemplateNoBoundVariables(
                 new LetExpr(node.tok, node.Dummies, newRhss.ToList(),

--- a/Source/Core/StandardVisitor.cs
+++ b/Source/Core/StandardVisitor.cs
@@ -311,9 +311,9 @@ namespace Microsoft.Boogie {
     public virtual Expr VisitLetExpr(LetExpr node) {
       Contract.Requires(node != null);
       Contract.Ensures(Contract.Result<Expr>() != null);
-      node.Body = this.VisitExpr(node.Body);
-      node.Dummies = this.VisitVariableSeq(node.Dummies);
       node.Rhss = this.VisitExprSeq(node.Rhss);
+      node.Dummies = this.VisitVariableSeq(node.Dummies);
+      node.Body = this.VisitExpr(node.Body);
       return node;
     }
     public virtual Formal VisitFormal(Formal node) {
@@ -878,9 +878,9 @@ namespace Microsoft.Boogie {
           return this.VisitBinderExpr(node);
       }
       public override Expr VisitLetExpr(LetExpr node) {
-          this.VisitExpr(node.Body);
-          this.VisitVariableSeq(node.Dummies);
           this.VisitExprSeq(node.Rhss);
+          this.VisitVariableSeq(node.Dummies);
+          this.VisitExpr(node.Body);
           return node;
       }
       public override Formal VisitFormal(Formal node)

--- a/Test/letexpr/git-issue-192.bpl
+++ b/Test/letexpr/git-issue-192.bpl
@@ -1,0 +1,24 @@
+// RUN: %boogie "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+type AAA;
+type BBB;
+type CCC;
+
+function Apple(AAA, BBB): CCC;
+function Banana(int): BBB;
+
+procedure Proc()
+{
+  var m: [AAA]CCC;
+  m :=
+      // Once upon a time, the lambda lifting for the following body switched
+      // the roles of 15 and Banana(700), by traversing the components of let
+      // expressions in a different order when figuring out holes from when
+      // replacing the holes with new bound variables of the lambda lifting.
+      (lambda aaa: AAA :: 
+          (var g := 15;
+             Apple(aaa, Banana(700))
+          )
+      );
+}

--- a/Test/letexpr/git-issue-192.bpl.expect
+++ b/Test/letexpr/git-issue-192.bpl.expect
@@ -1,0 +1,2 @@
+
+Boogie program verifier finished with 1 verified, 0 errors

